### PR TITLE
Update create and update action for reports controller [SCI-5641]

### PR DESCRIPTION
**Pull request by aignatov-bio**
Jira ticket: [SCI-5641](https://biosistemika.atlassian.net/browse/SCI-5641)

### What was done
Update create and update action for reports controller

### CODE REVIEW

File: app/views/reports/new.html.erb [hunk 1]
```diff
@@ -65,13 +65,13 @@
           </button>
           <ul id="reportGenerateMenuDropdown" class="dropdown-menu dropdown-menu-right" aria-labelledby="reportGenerateMenu">
             <li>
-              <%= link_to '#', remote: true, id: 'createNewReport' do %>
+              <%= link_to reports_path, remote: true, id: 'saveAsNewReport', data: {create_url: reports_path} do %>
```
Comment (okriuchykhin) [inline app/views/reports/new.html.erb:68]: spaces are missing after brackets

File: app/views/reports/new.html.erb [hunk 2]
```diff
@@ -65,13 +65,13 @@
...
                 <i class="fas fa-plus-circle"></i>
                 <%= t("projects.reports.new.save_as_new_report") %>
               <% end %>
             </li>
             <li>
-              <%= link_to '#', remote: true, id: 'UpdateReport' do %>
+              <%= link_to '#', remote: true, id: 'UpdateReport', data: {update_url: report_path(@report)}  do %>
```
Comment (okriuchykhin) [inline app/views/reports/new.html.erb:74]: same

File: app/components/reports/checkbox_input_component.html.erb [hunk 3]
```diff
@@ -1,6 +1,6 @@
 <div>
   <div class="sci-checkbox-container">
-    <%= check_box_tag @name, true, false, class: 'sci-checkbox' %>
+    <%= check_box_tag @name, true, (@value && @value == 'true'), class: 'sci-checkbox', data: {type: 'CheckboxInputComponent'} %>
```
Comment (okriuchykhin) [inline app/components/reports/checkbox_input_component.html.erb:3]: Isn't @value == 'true' just enough here?

Comment (okriuchykhin) [reply app/components/reports/checkbox_input_component.html.erb:3]: spaces are missing

Comment (aignatov-bio) [reply app/components/reports/checkbox_input_component.html.erb:3]: For the new report, @value is nil. It will generate an error.

Comment (okriuchykhin) [reply app/components/reports/checkbox_input_component.html.erb:3]: nil == 'true' will produce error?

Comment (aignatov-bio) [reply app/components/reports/checkbox_input_component.html.erb:3]: I will check again

File: app/controllers/reports_controller.rb [hunk 4]
```diff
@@ -551,4 +525,16 @@ def log_activity(type_of, report = @report)
...
+  def prepare_report_content
+    if params[:project_content]
+      @report_contents = ReportActions::ReportContent
+                         .new(params[:project_content], params[:settings], current_user)
+                         .generate_content
+    end
+  end
+
+  def generate_pdf_report
+    Reports::PdfJob.perform_later(@report, 'template_1', current_user) if @report.id
```
Comment (okriuchykhin) [inline app/controllers/reports_controller.rb:538]: may be better to check if report was saved? '@report.persisted?'

File: app/services/report_actions/report_content.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+module ReportActions
+  class ReportContent
+    include Canaid::Helpers::PermissionsHelper
+
+    MY_MODULE_ADDITIONAL_ELEMENTS = []
+
+    def initialize(content, setting, user)
+      @content = content
+      @setting = setting
```
Comment (okriuchykhin) [inline app/services/report_actions/report_content.rb:11]: shouldn't it be plural? 'settings'

File: app/controllers/reports_controller.rb [hunk 6]
```diff
@@ -117,30 +106,15 @@ def edit
...
-      report_contents = JSON.parse(params.delete(:report_contents))
-    rescue
-      continue = false
-    end
-
     @report.last_modified_by = current_user
     @report.assign_attributes(report_params)
 
-    if continue && @report.save_with_contents(report_contents)
+    if @report_contents && @report.save_with_contents(@report_contents, params[:template_values])
```
Comment (okriuchykhin) [inline app/controllers/reports_controller.rb:112]: Move 'save_with_contents' logic to the service

File: app/services/report_actions/report_content.rb [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        children << {
+          'type_of' => 'my_module_repository',
+          'id' => {
+            'repository_id' => repository.id,
+            'my_module_id' => my_module.id
+          }
+        }
+      end
+
+      MY_MODULE_ADDITIONAL_ELEMENTS.each do |e|
```
Comment (okriuchykhin) [inline app/services/report_actions/report_content.rb:84]: Use more descriptive variable name

File: app/services/report_actions/save_report.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
```
Comment (okriuchykhin) [inline app/services/report_actions/save_report.rb:1]: Why separate service?? You can simply create report elements within report_content.rb service.

File: app/assets/javascripts/reports/new.js [hunk 9]
```diff
@@ -1034,6 +1048,25 @@ function reportHandsonTableConverter() {
         // TODO
       });
     });
+
+    $('.reports-new').on('click', '#saveAsNewReport', function(e) {
+      var params = getReportData();
+      params.report.name = 'New ' + params.report.name;
+      e.preventDefault();
+      $.post(this.dataset.createUrl, params, function() {
+        // TODO
```
Comment (okriuchykhin) [inline app/assets/javascripts/reports/new.js:1057]: Shouldn't it be implemented?

Comment (aignatov-bio) [reply app/assets/javascripts/reports/new.js:1057]: We have a ticket for front-end validations (check report name and selected content). I think it can be done in the scope of this ticket.

Comment (okriuchykhin) [reply app/assets/javascripts/reports/new.js:1057]: Ok

File: app/models/report.rb [hunk 10]
```diff
@@ -32,24 +32,24 @@ class Report < ApplicationRecord
   has_many :report_elements, inverse_of: :report, dependent: :delete_all
 
   DEFAULT_SETTINGS = {
-    all_tasks: true,
+    all_tasks: 'true',
```
Comment (okriuchykhin) [inline app/models/report.rb:35]: Why did you change all booleans to strings?

Comment (aignatov-bio) [reply app/models/report.rb:35]: When we save it to the database it will become a string. And we need to compare it with `'true'` string.

Comment (okriuchykhin) [reply app/models/report.rb:35]: It doesn't

Comment (aignatov-bio) [reply app/models/report.rb:35]: When you create a new report it is ok. But when you load Report:
```
[1] pry(main)> Report.last.settings
  Report Load (1.7ms)  SELECT "reports".* FROM "reports" ORDER BY "reports"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> {"all_tasks"=>"true",
 "task"=>
  {"protocol"=>{"description"=>"true", "completed_steps"=>"true", "uncompleted_steps"=>"true", "step_checklists"=>"true", "step_files"=>"true", "step_tables"=>"true", "step_comments"=>"true"},
   "file_results"=>"false",
   "file_results_previews"=>"false",
   "table_results"=>"true",
   "text_results"=>"true",
   "result_comments"=>"true",
   "result_order"=>"atoz",
   "activities"=>"true"}}
```

Comment (okriuchykhin) [reply app/models/report.rb:35]: Report.last.settings
  Report Load (0.7ms)  SELECT "reports".* FROM "reports" ORDER BY "reports"."id" DESC LIMIT $1  [["LIMIT", 1]]
=> {"all_tasks"=>true,
 "task"=>
  {"protocol"=>{"description"=>true, "completed_steps"=>true, "uncompleted_steps"=>true, "step_checklists"=>true, "step_files"=>true, "step_tables"=>true, "step_comments"=>true},
   "file_results"=>false,
   "file_results_previews"=>false,
   "table_results"=>true,
   "text_results"=>true,
   "result_comments"=>true,
   "result_order"=>"atoz",
   "activities"=>true}}

Comment (aignatov-bio) [reply app/models/report.rb:35]: Looks like an issue in params. In params, it is a string.

File: app/services/report_actions/report_content.rb [hunk 11]
```diff
@@ -0,0 +1,157 @@
+# frozen_string_literal: true
+
+module ReportActions
+  class ReportContent
+    include Canaid::Helpers::PermissionsHelper
+
+    MY_MODULE_ADDONS_ELEMENTS = [].freeze
```
Comment (okriuchykhin) [inline app/services/report_actions/report_content.rb:7]: If it's frozen how do you plan to extend it?

File: app/models/report.rb (hunk 10 above)
Comment (aignatov-bio) [reply app/models/report.rb:35]: So we need or convert params to boolean. Or change here to string.

File: app/services/report_actions/report_content.rb (hunk 11 above)
Comment (aignatov-bio) [reply app/services/report_actions/report_content.rb:7]: Oh probably it is rubocop, I will fix it.

**Review by okriuchykhin — CHANGES_REQUESTED** [review]
Fix booleans in settings JSON

**Review by okriuchykhin — APPROVED** [review]
